File: ts/project/.sdk/tm/rust/utility/feature_add.rs
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::helpers::get_str;
use crate::core::types::FeatureRef;

// feature_add_util appends a feature to the client's feature list. A
// feature whose add_options() names an already-added feature with
// "__before__", "__after__" or "__replace__" positions itself relative to
// it — mirroring the ts featureAdd. The first match wins; when no ordering
// option matches, the feature is appended.
pub fn feature_add_util(ctx: &Rc<Context>, f: FeatureRef) {
    let client = match ctx.client.borrow().clone() {
        Some(c) => c,
        None => return,
    };

    let fopts = f.borrow().add_options();

    if let Some(fopts) = fopts {
        let before = get_str(&fopts, "__before__").unwrap_or_default();
        let after = get_str(&fopts, "__after__").unwrap_or_default();
        let replace = get_str(&fopts, "__replace__").unwrap_or_default();

        if !before.is_empty() || !after.is_empty() || !replace.is_empty() {
            let mut features = client.features.borrow_mut();
            for i in 0..features.len() {
                let name = features[i].borrow().name();
                if before == name {
                    features.insert(i, f);
                    return;
                }
                if after == name {
                    features.insert(i + 1, f);
                    return;
                }
                if replace == name {
                    features[i] = f;
                    return;
                }
            }
        }
    }

    client.features.borrow_mut().push(f);
}

File: ts/project/.sdk/tm/go/utility/feature_add.go
package utility

import "GOMODULE/core"

func featureAddUtil(ctx *core.Context, f core.Feature) {
	client := ctx.Client
	features := client.Features

	var fopts map[string]any
	if af, ok := f.(interface{ AddOptions() map[string]any }); ok {
		fopts = af.AddOptions()
	}

	if fopts != nil {
		before, _ := fopts["__before__"].(string)
		after, _ := fopts["__after__"].(string)
		replace, _ := fopts["__replace__"].(string)

		if "" != before || "" != after || "" != replace {
			for i, ef := range features {
				name := ef.GetName()
				if before == name {
					client.Features = append(features[:i],
						append([]core.Feature{f}, features[i:]...)...)
					return
				}
				if after == name {
					client.Features = append(features[:i+1],
						append([]core.Feature{f}, features[i+1:]...)...)
					return
				}
				if replace == name {
					features[i] = f
					client.Features = features
					return
				}
			}
		}
	}

	client.Features = append(features, f)
}
